### PR TITLE
micrometer-api module split off from micrometer-core

### PR DESCRIPTION
**Pull request by shakuzen**
This moves a portion of existing classes in micrometer-core to a new micrometer-api module. micrometer-core depends on the API module and now micrometer-core contains only the instrumentation that we maintain.

Projects that want to write their own instrumentation now only need to depend on the micrometer-api module. This will save some kilobytes in the JAR size for them.

End users may continue to depend on micrometer-core to get the instrumentation we provide in addition to the API.

Package names are changed, which will require users to update their imports. We could not keep the same package name without causing split packages across micrometer-api and micrometer-core.

### CODE REVIEW

**Review by shakuzen — COMMENTED** [review]
I know this is a huge changeset with mostly uninteresting noise (rename of package in imports), so I've commented on most files that weren't that.

**Comment by shakuzen** [comment]
Registry implementations continue to depend on `micrometer-core`, but should probably depend on `micrometer-api`. There are some issues with that change in the current code, so I left it as something to consider separately from this change.

File: dependencies.gradle [hunk 1]
```diff
@@ -38,8 +38,12 @@ def VERSIONS = [
         'info.ganglia.gmetric4j:gmetric4j:latest.release',
         'io.prometheus:simpleclient_common:latest.release',
         'io.prometheus:simpleclient_pushgateway:latest.release',
+        // TODO support HttpTagsProvider variants based on Jakarta APIs?
+//        'jakarta.servlet:jakarta.servlet-api:latest.release',
```
Comment (shakuzen) [inline dependencies.gradle:42]: In working on this, I noticed we have a javax servlet based http tags provider. We should probably also have jakarta-based one. Supporting either in the instrumentation may be a challenge. But something to work on separately from this PR.

**Comment by shakuzen** [comment]
On JAR size reduction due to this splitting,
micrometer-core 2.0.0.M1: 664 KB

With these changes,
micrometer-api: 393 KB
micrometer-core: 279 KB

File: micrometer-core/build.gradle [hunk 2]
```diff
@@ -50,15 +37,14 @@ dependencies {
     optionalApi 'org.apache.kafka:kafka-clients'
     optionalApi 'org.apache.kafka:kafka-streams'
 
+    testImplementation project(':micrometer-test')
```
[the remark this answers is not in the document]
Comment (jonatan-ivanov) [reply micrometer-core/build.gradle:41]: Move Issue to API?

**Review by jonatan-ivanov — APPROVED** [review]

**Comment by kazuki43zoo** [comment]
@shakuzen This change does not apply to `main` branch. Is this intentional?

**Comment by shakuzen** [comment]
@kazuki43zoo This change was eventually reverted in #3046, which will be part of today's release of `2.0.0-M3`. Thanks for asking. It made me realize #3046 wasn't in the milestone, and didn't link to the related issues, which would have helped discoverability. We plan to mention this prominently in the release notes, but we should have left a comment on this pull request to make it easier for people to figure out what happened. Hopefully this clears up what happened.

**Pull request merged**